drivers/sensor/mhz19b/mhz19b.c

```diff
@@ -0,0 +1,328 @@
+	    chan != SENSOR_CHAN_CURRENT) {
+		return -ENOTSUP;
+	}
+
+	return mhz19b_poll_data(dev, MHZ19B_CMD_GET_CO2);
+}
+
+static const struct sensor_driver_api mhz19b_api_funcs = {
+	.attr_set = mhz19b_attr_set,
+	.attr_get = mhz19b_attr_get,
+	.trigger_set = mhz19b_trigger_set,
+	.sample_fetch = mhz19b_sample_fetch,
+	.channel_get = mhz19b_channel_get,
+};
+
+static int mhz19b_init(const struct device *dev)
+{
+	struct mhz19b_data *data = dev->data;
+	const struct mhz19b_cfg *cfg = dev->config;
+	struct sensor_value val;
+	int ret;
+
+	uart_irq_rx_disable(data->uart_dev);
+	uart_irq_tx_disable(data->uart_dev);
+
+	mhz19b_uart_flush(data->uart_dev);
+
+	uart_irq_callback_set(data->uart_dev, cfg->cb);
+
+	k_sem_init(&data->rx_sem, 0, 1);
+	k_sem_init(&data->tx_sem, 1, 1);
+
+	/* Configure default detection range */
+	val.val1 = cfg->range;
+
+	ret = mhz19b_attr_set(dev, SENSOR_CHAN_CO2, SENSOR_ATTR_FULL_SCALE, &val);
+	if (ret != 0) {
+		LOG_ERR("Error setting default range %d", val.val1);
+		return ret;
+	}
+
+	/* Configure ABC logic */
+	val.val1 = (uint32_t)cfg->abc_on;
+	ret = mhz19b_attr_set(dev, SENSOR_CHAN_CO2, SENSOR_ATTR_MHZ19B_ABC, &val);
+	if (ret != 0) {
+		LOG_ERR("Error setting default ABC %s", val.val1 == 1 ? "on" : "off");
+	}
+
+	return ret;
+}
+
+#define MHZ19B_UART_ISR_FN(inst)								\
+	static void mhz19b_uart_isr_##inst(const struct device *uart_dev, void *user_data)	\
+	{											\
+		const struct device *dev = DEVICE_DT_INST_GET(inst);				\
+		struct mhz19b_data *data = dev->data;						\
+		static int xfer_bytes;								\
+												\
+		ARG_UNUSED(user_data);								\
+												\
+		if (uart_dev == NULL) {								\
+			return;									\
+		}										\
+												\
+		if (!uart_irq_update(uart_dev)) {						\
+			return;									\
+		}										\
+												\
+		if (uart_irq_rx_ready(uart_dev)) {						\
+			xfer_bytes += uart_fifo_read(uart_dev, &data->rd_data[xfer_bytes],	\
+						     MHZ19B_BUF_LEN - xfer_bytes);		\
+												\
+			if (xfer_bytes == MHZ19B_BUF_LEN) {					\
+				xfer_bytes = 0;							\
+				uart_irq_rx_disable(uart_dev);					\
+				k_sem_give(&data->rx_sem);					\
+				if (data->has_rsp) {						\
+					k_sem_give(&data->tx_sem);				\
+				}								\
+			}									\
+		}										\
+												\
+		if (uart_irq_tx_ready(uart_dev)) {						\
+			xfer_bytes +=								\
+				uart_fifo_fill(uart_dev, &mhz19b_cmds[data->cmd][xfer_bytes],	\
+					       MHZ19B_BUF_LEN - xfer_bytes);			\
+												\
+			if (xfer_bytes == MHZ19B_BUF_LEN) {					\
+				xfer_bytes = 0;							\
+				uart_irq_tx_disable(uart_dev);					\
+				if (!data->has_rsp) {						\
+					k_sem_give(&data->tx_sem);				\
+				}								\
+			}									\
+		}										\
+	}
+
+#define MHZ19B_INIT(inst)									\
+	MHZ19B_UART_ISR_FN(inst);								\
+												\
+	static struct mhz19b_data mhz19b_data_##inst = {					\
+		.uart_dev = DEVICE_DT_GET(DT_INST_BUS(inst)),					\
+	};											\
```

Comment:
Why isn't this in the config structure? It seems like wasted RAM to put it here.

Reply:
Done